charms/istio-pilot/lib/charms/istio_pilot/v0/ingress.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""# Ingress Interface Library.
+"""
+
+import logging
+from functools import cached_property
+
+from ops.charm import CharmBase
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import BlockedStatus, WaitingStatus
+from serialized_data_interface import get_interface, NoCompatibleVersions, NoVersionsListed
+
+logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "b521889515b34432b952f75c21e96dfc"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class IngressProviderAvailableEvent(EventBase):
+    """Event triggered when the ingress provider is ready for requests."""
+
+
+class IngressReadyEvent(EventBase):
+    """Event triggered when the ingress provider has returned the requested URL(s)."""
+
+
+class IngressFailedEvent(EventBase):
+    """Event triggered when something went wrong with the ingress relation."""
+
+
+class IngressRequirerEvents(ObjectEvents):
+    available = EventSource(IngressProviderAvailableEvent)
+    ready = EventSource(IngressReadyEvent)
+    failed = EventSource(IngressFailedEvent)
+
+
+class IngressRequirer(Object):
+    on = IngressRequirerEvents()
+
+    def __init__(self, charm: CharmBase, relation_endpoint: str):
```

Comment:
I think we should use the same pattern with the default relation name that has proven to be very successful for the O11y team: https://github.com/canonical/prometheus-k8s-operator/blob/37a82abd1909e59aa5ac0ff810cca74864ca13d4/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L835

Reply:
Also, notice the logic to validate the role of the relation, it saves lives: https://github.com/canonical/prometheus-k8s-operator/blob/37a82abd1909e59aa5ac0ff810cca74864ca13d4/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L857

Reply:
Last but not list, I cannot imagine _any scenario_ in which a requiring charm may want a relation with `limit` value different than `1` (why would you connect to _multiple ingress controllers_ at once?). We should probably enforce that.

Reply:
> why would you connect to multiple ingress controllers at once?

Well, actually, I could see a possible use-case where you want to SaaS a service to multiple separate networks and want to maximize isolation between those networks so you have multiple ingress controllers, each of which only has a network connection to one of the external networks. It's a pretty out-there example, but I don't know that it's worth putting in extra effort to add in an artificial limitation. Then again, knowing which ingress address to advertise to a particular related app would run up against the issue that Vault did where Juju doesn't provide enough info about a relation to know which address to present to it and the charm has to sort of guess based on things like subnet overlap. *shrug*

Reply:
But then how do you map those unit/app URIs to what you put in which configs. It sounds theoretically possible, but I do not see it as a problem worth solving atm :-)

Reply:
It has, in fact, come up with Vault when they needed different addresses sent to different client relations based on what network segment they are in (internal vs external), so it's not theoretical in the slightest. However, sorting that out properly requires some changes to how Juju presents info about a given relation to the charm, so it's currently only partially supported in Vault today (they use a heuristic based on subnets, which fails in many environments).

My point was just that it would take additional effort to try to enforce the limit (esp. when Juju already does that if you put it in), particularly given https://github.com/canonical/operator/issues/498, so I'm disinclined to put effort into that right now.

Reply:
The reason why we tend to enforce these limits in the logic of the library is that charm authors seldom think to think that aspect through. The default behavior of `limit` being "Pile on as many relations instances as you please" in my experience plays absolute havoc on the mental model of charm authors that predominantly seem to assume that "I have _up to one_ relation of this kind".

Reply:
It would be nice if the library could have a build-time hook to validate the `limit` field before the charm is even built.

Reply:
Added all of this.